python/ray/tests/test_global_state.py

```diff
@@ -17,7 +17,7 @@
 @pytest.mark.skipif(
     pytest_timeout is None,
     reason="Timeout package not installed; skipping test that may hang.")
-@pytest.mark.timeout(10)
+@pytest.mark.timeout(30)
```

Comment:
The test code itself doesn't take much time ( < 1s). The real issue is that `pytest.mark.timeout` considers the overall duration of test session for timing out. See the output below,

```
Test session starts (platform: win32, Python 3.8.11, pytest 5.4.3, pytest-sugar 0.9.4)
cachedir: .pytest_cache
rootdir: C:\Users\gagan\ray_project\ray\python
plugins: anyio-3.3.2, asyncio-0.15.1, lazy-fixture-0.6.3, repeat-0.9.1, rerunfailures-10.2, sugar-0.9.4, timeout-1.4.2
collecting ... DEBUG time_taken  0.35937976837158203 # Acutal time to complete the tests.

 ray\tests\test_global_state.py::test_replenish_resources ✓                       100% ██████████

Results (9.20s): # 9.20 s is the duration of the test session. Pytest will timeout if it exceeds 10s even though the actual duration for completing the test assertions is 0.35s.
       1 passed

```

May be we need to manually timeout the block of code instead of using pytest?

Reply:
That sounds good to me, do you know a good way to do this?

Reply:
https://stackoverflow.com/a/2282656 - This answer has some ideas which I think I can manually implement and try making it work with tests. Let me know if that's fine.

Reply:
Ah! That's UNIX only though.

Reply:
https://dreamix.eu/blog/webothers/timeout-function-in-python-3 - This uses Python threads. So, we can write a function inside our test function. Then call it from a thread and pass timeout value to `join`.

```python
def test_replenish_resources():
    def test_replenish_resources_sub():
        ...do stuff here
    
    test_thread = Thread(target=test_replenish_resources_sub)
    test_thread.start()
    test_thread.join(timeout=10)
```

Are threads in any way receive lesser memory in python?

Reply:
Also maybe the current solution is good enough for the time being, or is it actively causing problems/timing out?

Reply:
It's passing on my Azure VM consistently. Though on Github CI, passing of tests depends on how fast pytest starts up and runs tests irrespective of what's there inside the test. So, may be 30s is big enough to accommodate everything. We can rely on that for now. If it fails too many times on Github then we can think of other ways of timing out tests. Thoughts?

Reply:
Yes, that sounds great :)